src/openforms/formio/datastructures.py

```diff
@@ -23,22 +43,44 @@ class FormioConfigurationWrapper:
     formio configuration.
     """
 
-    _configuration: JSONObject
+    _configuration: FormioConfiguration
     # depth-first ordered of all components in the formio configuration tree
     _cached_component_map: dict[str, Component] | None = None
     _flattened_by_path: None | dict[str, Component] = None
     _reverse_flattened: None | dict[str, str] = None
 
-    def __init__(self, configuration: JSONObject):
+    def __init__(self, configuration: FormioConfiguration):
         self._configuration = configuration
 
     @property
     def component_map(self) -> dict[str, Component]:
         if self._cached_component_map is None:
-            self._cached_component_map = {
-                component["key"]: component
-                for component in iter_components(self.configuration, recursive=True)
-            }
+            self._cached_component_map = {}
+
+            for component in iter_components(self.configuration, recursive=True):
+                # first, ensure we add every component by its own key so that we can
+                # look it up. This is okay *because* in our UI we enforce unique keys
+                # across the entire form, even if the key is present in an edit grid
+                # (repeating group). Note that formio is perfectly fine with a root
+                # 'foo' key and 'foo' key inside an editgrid. Our behaviour is different
+                # from that because of historical reasons...
+                self._cached_component_map[component["key"]] = component
+
+                # now, formio itself addresses components inside an edit grid with the
+                # pattern ``<editGridKey>.<componentKey>`` (e.g. in simple conditionals),
+                # which means that we also need to add the editgrid components to our
+                # 'registry' for easy lookups. See GH issue #4247 for one possible way
+                # this can cause crashes. So, we add the nested components with a
+                # namespaced key too.
+                #
+                # NOTE - this could conflict with a component outside the editgrid with
+                # this specific, explicit key. At the time of writing, this crashes on
+                # Formio's own demo site because it can't properly resolve the component,
+                # so we do not need to consider this case (it's broken anyway).
+                if component["type"] == "editgrid":
+                    editgrid_components = _get_editgrid_component_map(component)  # type: ignore
+                    self._cached_component_map.update(editgrid_components)
```

Comment:
This will add duplicate components in the returned mapping, but guessing it will be fine as long as it is only used for lookups, and not when iterating over `.items()/.keys()/.values()`

Reply:
Just fixed a bug related to that in 49f2e9c525f4f207d746b786af89a7c3817256cb, let's see if other tests break because of that. `.keys` and `.values` are not intended to be used outside of this data structure, but it's  blurry line :/